zerver/views/streams.py

```diff
@@ -315,12 +322,21 @@ def update_stream_backend(
         else:
             raise JsonableError(_("Invalid parameters"))
 
-    if is_private is not None:
-        # Default streams cannot be made private.
-        default_stream_ids = get_default_stream_ids_for_realm(stream.realm_id)
-        if is_private and stream.id in default_stream_ids:
+    # Ensure that a stream cannot be both a default stream for new users and private,
+    # and disallow making default streams private or private streams default.
+    if proposed_is_private and proposed_is_default_stream:
+        if is_default_stream and is_private:
+            raise JsonableError(
+                _("A stream cannot be both a default stream for new users and private.")
+            )
+
+        if is_private and proposed_is_default_stream:
             raise JsonableError(_("Default streams cannot be made private."))
 
+        if is_default_stream and proposed_is_private:
+            raise JsonableError(_("Private streams cannot be made default."))
```

Comment:
I think we don't need the three different error messages; we can just have it be "A default stream cannot be private." for all cases here. (This will save unnecessary work for translators)